R-package/R/lgb.Dataset.R

```diff
@@ -533,22 +533,47 @@ Dataset <- R6::R6Class(
 
     # Update parameters
     update_params = function(params) {
-
-      # Parameter updating
-      if (!lgb.is.null.handle(private$handle)) {
-        lgb.call(
-          "LGBM_DatasetUpdateParam_R"
-          , ret = NULL
-          , private$handle
+      if (lgb.is.null.handle(private$handle)) {
+        private$params <- modifyList(private$params, params)
+      } else {
+        call_state <- .Call(
+          "LGBM_DatasetUpdateParamWarning_R"
+          , lgb.params2str(private$params)
           , lgb.params2str(params)
+          , call_state
+          , PACKAGE = "lib_lightgbm"
         )
-        return(invisible(self))
+        call_state <- as.integer(call_state)
+        if (call_state != 0L) {
+
+          # raise error if raw data is freed
+          if (is.null(private$raw_data)) {
+            lgb.last_error()
+          }
+
+          # Overwrite paramms
+          private$params <- modifyList(private$params, params)
+          self$finalize()
+        }
       }
-      private$params <- modifyList(private$params, params)
       return(invisible(self))
 
     },
 
+    get_params = function() {
+      keys <- c("has_header", "header", "max_bin", "max_bin_by_feature", "bin_construct_sample_cnt", "subsample_for_bin", "min_data_in_bin", "use_missing",
+        "zero_as_missing", "sparse_threshold", "categorical_feature", "cat_feature", "categorical_column", "cat_column", 
+        "feature_pre_filter", "pre_partition", "is_pre_partition", "enable_bundle", "is_enable_bundle", "bundle", "max_conflict_rate ",
+        "is_enable_sparse", "is_sparse", "enable_sparse", "sparse", "forcedbins_filename")
+      ret <- list()
+      for (key in names(private$params)) {
```

Comment:
could you name this iteration variable `param_key` or something? `key()` is a function exported from `data.table` (one of our dependencies) so we should avoid that name to prevent any weird things happening